File: app/src/main/java/org/andstatus/app/user/CachedUsersAndActors.java
package org.andstatus.app.user;

import android.database.Cursor;

import org.andstatus.app.actor.GroupType;
import org.andstatus.app.context.MyContext;
import org.andstatus.app.data.ActorSql;
import org.andstatus.app.data.DbUtils;
import org.andstatus.app.data.GroupMembership;
import org.andstatus.app.data.MyQuery;
import org.andstatus.app.database.table.ActorTable;
import org.andstatus.app.database.table.GroupMembersTable;
import org.andstatus.app.database.table.TimelineTable;
import org.andstatus.app.database.table.UserTable;
import org.andstatus.app.net.social.Actor;
import org.andstatus.app.origin.Origin;
import org.andstatus.app.util.CollectionsUtil;
import org.andstatus.app.util.MyLog;
import org.andstatus.app.util.TriState;

import java.util.Collection;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.function.Supplier;

import androidx.annotation.NonNull;

import static org.andstatus.app.util.RelativeTime.SOME_TIME_AGO;

public class CachedUsersAndActors {
    private final MyContext myContext;
    public final Map<Long, User> users = new ConcurrentHashMap<>();
    public final Map<Long, Actor> actors = new ConcurrentHashMap<>();
    public final Map<Long, User> myUsers = new ConcurrentHashMap<>();
    public final Map<Long, Actor> myActors = new ConcurrentHashMap<>();
    /** key - friendId, set of values - IDs of my actors  */
    public final Map<Long, Set<Long>> friendsOfMyActors = new ConcurrentHashMap<>();

    public static CachedUsersAndActors newEmpty(MyContext myContext) {
        return new CachedUsersAndActors(myContext);
    }

    private CachedUsersAndActors(MyContext myContext) {
        this.myContext = myContext;
    }

    public int size() {
        return myUsers.size();
    }

    public CachedUsersAndActors initialize() {
        initializeMyUsers();
        initializeFriendsOfMyActors();
        loadTimelineActors();
        MyLog.v(this, () -> "Users list initialized, "
                + myUsers.size() + " users, "
                + myActors.size() + " my actors, "
                + friendsOfMyActors.size() + " friends");
        return this;
    }

    private void initializeMyUsers() {
        users.clear();
        actors.clear();
        myUsers.clear();
        myActors.clear();
        final String sql = "SELECT " + ActorSql.select()
                + " FROM " + ActorSql.tables()
                + " WHERE " + UserTable.IS_MY + "=" + TriState.TRUE.id;
        final Function<Cursor, Actor> function = cursor -> Actor.fromCursor(myContext, cursor);
        MyQuery.get(myContext, sql, function).forEach(this::updateCache);
    }

    private void initializeFriendsOfMyActors() {
        final String FOLLOWER_ID = "followerId";
        friendsOfMyActors.clear();
        final String sql = "SELECT DISTINCT " + ActorSql.select()
                + ", friends." + ActorTable.PARENT_ACTOR_ID + " AS " + FOLLOWER_ID
                + " FROM (" + ActorSql.tables() + ")"
                + " INNER JOIN (" + GroupMembership.selectMemberIds(myActors.keySet(), GroupType.FRIENDS, true)
                + " ) as friends ON friends." + GroupMembersTable.MEMBER_ID +
                "=" + ActorTable.TABLE_NAME + "." + ActorTable._ID;

        final Function<Cursor, Void> function = cursor -> {
            Actor friend = Actor.fromCursor(myContext, cursor);
            Actor me = Actor.load(myContext, DbUtils.getLong(cursor, FOLLOWER_ID));
            friendsOfMyActors.compute(friend.actorId, CollectionsUtil.addValue(me.actorId));
            return null;
        };
        MyQuery.get(myContext, sql, function);
    }

    public Actor load(long actorId) {
        Actor actor = Actor.load(myContext, actorId, false, Actor::getEmpty);
        if (isMe(actor)) loadFriendsOfMy(actor);
        return actor;
    }

    public Actor reload(Actor actor) {
        return reload(actor.actorId);
    }

    public Actor reload(long actorId) {
        Actor reloaded = Actor.load(myContext, actorId, true, Actor::getEmpty);
        if (isMe(reloaded)) loadFriendsOfMy(reloaded);
        return reloaded;
    }

    private void loadFriendsOfMy(Actor actor) {
        friendsOfMyActors.entrySet().stream().filter( entry -> entry.getValue().contains(actor.actorId))
                .forEach(entry ->
                        friendsOfMyActors.compute(entry.getKey(), CollectionsUtil.removeValue(actor.actorId)));
        MyQuery.getGroupMemberIds(myContext, actor.actorId, GroupType.FRIENDS)
                .forEach(friendId -> friendsOfMyActors.compute(friendId, CollectionsUtil.addValue(actor.actorId))
        );
    }

    private void loadTimelineActors() {
        final String sql = "SELECT " + ActorSql.select()
                + " FROM " + ActorSql.tables()
                + " WHERE " + ActorTable.TABLE_NAME + "." + ActorTable._ID + " IN ("
                + " SELECT DISTINCT " + TimelineTable.ACTOR_ID
                + " FROM " + TimelineTable.TABLE_NAME
                + ")";
        final Function<Cursor, Actor> function = cursor -> Actor.fromCursor(myContext, cursor);
        MyQuery.get(myContext, sql, function);
    }

    public boolean containsMe(@NonNull Collection<Actor> actors) {
        return actors.stream().anyMatch(this::isMe);
    }

    @Override
    public String toString() {
        return "MyUsers{\n" + myUsers + "\nMy actors: " + myActors + "\nMy friends: " + friendsOfMyActors + '}';
    }

    public boolean isMeOrMyFriend(Actor actor) {
        return actor.nonEmpty() && (isMe(actor) || friendsOfMyActors.containsKey(actor.actorId));
    }

    public boolean isMe(@NonNull Actor actor) {
        return isMe(actor.actorId);
    }

    public boolean isMe(long actorId) {
        return actorId != 0 && (
            myActors.containsKey(actorId) || myUsers.values().stream().anyMatch(user -> user.actorIds.contains(actorId))
        );
    }

    public Actor lookupUser(Actor actor) {
        if (actor.isGroupDefinitely()) {
            return actor;
        }

        if (actor.user == User.EMPTY && actor.actorId != 0) {
            actor.user = User.load(myContext, actor.actorId);
        }
        if (actor.user == User.EMPTY && actor.isWebFingerIdValid()) {
            actor.user = User.load(myContext, MyQuery.webFingerIdToId(myContext, 0, actor.getWebFingerId()));
        }
        if (actor.user == User.EMPTY) {
            actor.user = User.getNew();
        } else {
            updateCache(actor);
        }
        return actor;
    }

    /** Tries to find this actor in this origin
     * Returns the same Actor, if not found */
    public Actor toOrigin(Actor actor, Origin origin) {
        return actor.origin.equals(origin)
                ? actor
                : actor.user.actorIds.stream().map(id -> actors.getOrDefault(id, Actor.EMPTY))
                .filter(a -> a != Actor.EMPTY && a.origin.equals(origin))
                .findAny().orElse(actor);
    }

    @NonNull
    public User userFromActorId(long actorId, Supplier<User> userSupplier) {
        if (actorId == 0) return User.EMPTY;
        final User user1 = actors.getOrDefault(actorId, Actor.EMPTY).user;
        return user1.nonEmpty()
                ? user1
                : users.values().stream().filter(user -> user.actorIds.contains(actorId)).findFirst().orElseGet(userSupplier);
    }

    public void updateCache(@NonNull Actor actor) {
        if (actor.isEmpty()) return;

        final User user = actor.user;
        final long userId = user.userId;
        final long actorId = actor.actorId;
        if (actorId != 0) {
            user.actorIds.add(actorId);
            updateCachedActor(actors, actor);
            if (user.isMyUser().isTrue) updateCachedActor(myActors, actor);
        }
        if (userId == 0) return;

        User cached = users.getOrDefault(userId, User.EMPTY);
        if (cached.isEmpty()) {
            users.putIfAbsent(userId, user);
            if (user.isMyUser().isTrue) myUsers.putIfAbsent(userId, user);
        } else if (user.isMyUser().isTrue && cached.isMyUser().untrue) {
            user.actorIds.addAll(cached.actorIds);
            users.put(userId, user);
            myUsers.put(userId, user);
        } else if (actorId != 0) {
            cached.actorIds.add(actorId);
        }
    }

    private void updateCachedActor(Map<Long, Actor> actors, Actor actor) {
        if (actor.isEmpty()) return;

        if (actor.getUpdatedDate() <= SOME_TIME_AGO) {
            actors.putIfAbsent(actor.actorId, actor);
            return;
        }
        if (actor.isBetterToCacheThan(actors.get(actor.actorId))) {
            actors.put(actor.actorId, actor);
            myActors.computeIfPresent(actor.actorId, (id, actor1) -> actor);
        }
    }
}

File: app/src/main/java/org/andstatus/app/user/CachedUsersAndActors.kt
package org.andstatus.app.user

import android.database.Cursor
import android.provider.BaseColumns
import org.andstatus.app.actor.GroupType
import org.andstatus.app.context.MyContext
import org.andstatus.app.data.ActorSql
import org.andstatus.app.data.DbUtils
import org.andstatus.app.data.GroupMembership
import org.andstatus.app.data.MyQuery
import org.andstatus.app.database.table.ActorTable
import org.andstatus.app.database.table.GroupMembersTable
import org.andstatus.app.database.table.TimelineTable
import org.andstatus.app.database.table.UserTable
import org.andstatus.app.net.social.Actor
import org.andstatus.app.origin.Origin
import org.andstatus.app.util.CollectionsUtil
import org.andstatus.app.util.MyLog
import org.andstatus.app.util.RelativeTime
import org.andstatus.app.util.StopWatch
import org.andstatus.app.util.TriState
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Supplier

class CachedUsersAndActors private constructor(private val myContext: MyContext) {
    val users: MutableMap<Long, User> = ConcurrentHashMap()
    val actors: MutableMap<Long, Actor> = ConcurrentHashMap()
    val actorGroupTypes: MutableMap<Long, GroupType> = ConcurrentHashMap()
    val originIdAndUsernameToActorId: MutableMap<String, Long> = ConcurrentHashMap()
    val myUsers: MutableMap<Long, User> = ConcurrentHashMap()
    val myActors: MutableMap<Long, Actor> = ConcurrentHashMap()

    /** key - friendId, set of values - IDs of my actors   */
    val friendsOfMyActors: MutableMap<Long, MutableSet<Long>> = ConcurrentHashMap()
    val followersOfMyActors: MutableMap<Long, MutableSet<Long>> = ConcurrentHashMap()
    fun size(): Int {
        return myUsers.size
    }

    fun initialize(): CachedUsersAndActors {
        val stopWatch: StopWatch = StopWatch.createStarted()
        initializeMyUsers()
        initializeMyFriendsOrFollowers(GroupType.FRIENDS, friendsOfMyActors)
        initializeMyFriendsOrFollowers(GroupType.FOLLOWERS, followersOfMyActors)
        loadTimelineActors()
        MyLog.i(this, "usersInitializedMs:" + stopWatch.time + "; "
                + myUsers.size + " users, "
                + myActors.size + " my actors, "
                + friendsOfMyActors.size + " friends, "
                + followersOfMyActors.size + " followers")
        return this
    }

    private fun initializeMyUsers() {
        users.clear()
        actors.clear()
        myUsers.clear()
        myActors.clear()
        val sql = ("SELECT " + ActorSql.selectFullProjection()
                + " FROM " + ActorSql.allTables()
                + " WHERE " + UserTable.IS_MY + "=" + TriState.TRUE.id)
        val function = Function<Cursor, Actor> { cursor: Cursor -> Actor.fromCursor(myContext, cursor, true) }
        MyQuery.get(myContext, sql, function).forEach(Consumer { actor: Actor -> updateCache(actor) })
    }

    private fun initializeMyFriendsOrFollowers(groupType: GroupType, groupMembers: MutableMap<Long, MutableSet<Long>>) {
        val MY_ACTOR_ID = "myActorId"
        groupMembers.clear()
        val sql = ("SELECT DISTINCT " + ActorSql.selectFullProjection()
                + ", friends." + ActorTable.PARENT_ACTOR_ID + " AS " + MY_ACTOR_ID
                + " FROM (" + ActorSql.allTables() + ")"
                + " INNER JOIN (" + GroupMembership.selectMemberIds(myActors.keys, groupType, true)
                + " ) as friends ON friends." + GroupMembersTable.MEMBER_ID +
                "=" + ActorTable.TABLE_NAME + "." + BaseColumns._ID)
        val function = Function<Cursor, Void> { cursor: Cursor ->
            val other: Actor = Actor.fromCursor(myContext, cursor, true)
            val me: Actor = Actor.load(myContext, DbUtils.getLong(cursor, MY_ACTOR_ID))
            groupMembers.compute(other.actorId, CollectionsUtil.addValue(me.actorId))
            null
        }
        MyQuery.get(myContext, sql, function)
    }

    fun reload(actor: Actor): Actor {
        return load(actor.actorId, true)
    }

    @JvmOverloads
    fun load(actorId: Long, reloadFirst: Boolean = false): Actor {
        val actor: Actor = Actor.load(myContext, actorId, reloadFirst) { Actor.getEmpty() }
        if (reloadFirst && isMe(actor)) {
            reloadFriendsOrFollowersOfMy(GroupType.FRIENDS, friendsOfMyActors, actor)
            reloadFriendsOrFollowersOfMy(GroupType.FOLLOWERS, followersOfMyActors, actor)
        }
        return actor
    }

    private fun reloadFriendsOrFollowersOfMy(groupType: GroupType, groupMembers: MutableMap<Long, MutableSet<Long>>, actor: Actor) {
        groupMembers.entries.stream().filter { entry: MutableMap.MutableEntry<Long, MutableSet<Long>> -> entry.value.contains(actor.actorId) }
                .forEach { entry: MutableMap.MutableEntry<Long, MutableSet<Long>> -> groupMembers.compute(entry.key, CollectionsUtil.removeValue(actor.actorId)) }
        GroupMembership.getGroupMemberIds(myContext, actor.actorId, groupType)
                .forEach(Consumer { friendId: Long -> groupMembers.compute(friendId, CollectionsUtil.addValue(actor.actorId)) }
                )
    }

    private fun loadTimelineActors() {
        val sql = ("SELECT " + ActorSql.selectFullProjection()
                + " FROM " + ActorSql.allTables()
                + " WHERE " + ActorTable.TABLE_NAME + "." + BaseColumns._ID + " IN ("
                + " SELECT DISTINCT " + TimelineTable.ACTOR_ID
                + " FROM " + TimelineTable.TABLE_NAME
                + ")")
        val function = Function<Cursor, Actor> { cursor: Cursor -> Actor.fromCursor(myContext, cursor, true) }
        MyQuery.get(myContext, sql, function)
    }

    fun containsMe(actors: Collection<Actor>): Boolean {
        return actors.stream().anyMatch { actor: Actor -> this.isMe(actor) }
    }

    override fun toString(): String {
        return "MyUsers{\n$myUsers\nMy actors: $myActors\nMy friends: $friendsOfMyActors}"
    }

    fun isMeOrMyFriend(actor: Actor): Boolean {
        return actor.nonEmpty && (isMe(actor) || friendsOfMyActors.containsKey(actor.actorId))
    }

    fun isMe(actor: Actor): Boolean {
        return isMe(actor.actorId)
    }

    fun isMe(actorId: Long): Boolean {
        return actorId != 0L && (myActors.containsKey(actorId) || myUsers.values.stream().anyMatch { user: User -> user.actorIds.contains(actorId) })
    }

    fun lookupUser(actor: Actor): Actor {
        if (actor.user.nonEmpty) {
            updateCache(actor)
            return actor
        }
        var user2: User = User.EMPTY
        if (user2.isEmpty && actor.actorId != 0L) {
            user2 = User.load(myContext, actor.actorId)
        }
        if (user2.isEmpty && actor.isWebFingerIdValid()) {
            user2 = User.load(myContext, MyQuery.webFingerIdToId(myContext, 0, actor.getWebFingerId(), false))
        }
        if (user2.isEmpty) {
            user2 = User.getNew()
        }
        if (actor.user.isMyUser().known) {
            user2.setIsMyUser(actor.user.isMyUser())
        }
        actor.user = user2
        if (actor.user.nonEmpty) {
            updateCache(actor)
        }
        return actor
    }

    /** Tries to find this actor in this origin
     * Returns the same Actor, if not found  */
    fun toOrigin(actor: Actor, origin: Origin): Actor {
        return if (actor.origin == origin) actor else actor.user.actorIds.stream()
                .map { id: Long -> actors.getOrDefault(id, Actor.EMPTY) }
                .filter { a: Actor -> a !== Actor.EMPTY && a.origin == origin }
                .findAny().orElse(actor)
    }

    fun userFromActorId(actorId: Long, userSupplier: Supplier<User>): User {
        if (actorId == 0L) return User.EMPTY
        val user1 = actors.getOrDefault(actorId, Actor.EMPTY).user
        return if (user1.nonEmpty) user1 else users.values.stream()
                .filter { user: User -> user.actorIds.contains(actorId) }
                .findFirst()
                .orElseGet(userSupplier)
    }

    fun idToGroupType(actorId: Long): GroupType {
        val groupTypeCached = actorGroupTypes.get(actorId)
        if (groupTypeCached != null) return groupTypeCached
        val groupTypeStored: GroupType = GroupType.fromId(MyQuery.idToLongColumnValue(
                myContext.getDatabase(), ActorTable.TABLE_NAME, ActorTable.GROUP_TYPE, actorId))
        actorGroupTypes[actorId] = groupTypeStored
        return groupTypeStored
    }

    fun updateCache(actor: Actor) {
        if (actor.isEmpty) return
        val user = actor.user
        val userId = user.userId
        val actorId = actor.actorId
        if (actorId != 0L) {
            if (userId != 0L) {
                user.actorIds.add(actorId)
            }
            updateCachedActor(actors, actor)
            if (user.isMyUser().isTrue) updateCachedActor(myActors, actor)
        }
        if (userId == 0L) return
        val cached = users.getOrDefault(userId, User.EMPTY)
        if (cached.isEmpty) {
            users.putIfAbsent(userId, user)
            if (user.isMyUser().isTrue) myUsers.putIfAbsent(userId, user)
        } else if (user.isMyUser().isTrue && cached.isMyUser().untrue) {
            user.actorIds.addAll(cached.actorIds)
            users[userId] = user
            myUsers[userId] = user
        } else if (actorId != 0L) {
            cached.actorIds.add(actorId)
        }
    }

    private fun updateCachedActor(actors: MutableMap<Long, Actor>, actor: Actor) {
        if (actor.isEmpty) return
        if (actor.getUpdatedDate() <= RelativeTime.SOME_TIME_AGO) {
            actors.putIfAbsent(actor.actorId, actor)
            return
        }
        if (actor.isBetterToCacheThan(actors.get(actor.actorId))) {
            actors[actor.actorId] = actor
            actorGroupTypes[actor.actorId] = actor.groupType
            if (actor.isOidReal()) {
                originIdAndUsernameToActorId[actor.origin.id.toString() + ";" + actor.getUsername()] = actor.actorId
            }
            myActors.computeIfPresent(actor.actorId, { id: Long, actor1: Actor -> actor })
        }
    }

    companion object {
        fun newEmpty(myContext: MyContext): CachedUsersAndActors {
            return CachedUsersAndActors(myContext)
        }
    }
}